hazelcast/src/main/java/com/hazelcast/map/impl/operation/EntryOperation.java

```diff
@@ -248,76 +251,157 @@ private void throwModificationInReadOnlyException() {
+        }
+    }
+
+    private void unlock(Data finalDataKey, String finalCaller, long finalThreadId, long finalCallId, Throwable cause) {
+        boolean unlocked = recordStore.unlock(finalDataKey, finalCaller, finalThreadId, finalCallId);
+        if (!unlocked) {
+            throw new IllegalStateException(
+                    String.format("Could not unlock by the caller=%s and threadId=%d", finalCaller, threadId), cause);
+        }
+    }
+
+    @SuppressWarnings({"unchecked", "checkstyle:methodlength"})
     private void updateAndUnlock(Data previousValue, Data newValue, EntryEventType modificationType, String caller,
                                  long threadId, final Object result, long now) {
         EntryOffloadableSetUnlockOperation updateOperation = new EntryOffloadableSetUnlockOperation(name, modificationType,
                 dataKey, previousValue, newValue, caller, threadId, now, entryProcessor.getBackupProcessor());
 
         updateOperation.setPartitionId(getPartitionId());
-        ops.invokeOnPartition(updateOperation).andThen(new ExecutionCallback<Object>() {
+        updateOperation.setReplicaIndex(0);
+        updateOperation.setNodeEngine(getNodeEngine());
+        updateOperation.setCallerUuid(getCallerUuid());
+        OperationAccessor.setCallerAddress(updateOperation, getCallerAddress());
+        @SuppressWarnings("checkstyle:anoninnerlength")
+        OperationResponseHandler setUnlockResponseHandler = new OperationResponseHandler() {
             @Override
-            public void onResponse(Object response) {
-                try {
-                    getOperationResponseHandler().sendResponse(EntryOperation.this, result);
-                } finally {
-                    ops.onCompletionAsyncOperation(EntryOperation.this);
+            public void sendResponse(Operation op, Object response) {
+                if (isRetryable(response) || isTimeout(response)) {
+                    retry(op);
+                } else {
+                    handleResponse(response);
                 }
             }
 
-            @Override
-            public void onFailure(Throwable t) {
-                try {
-                    // EntryOffloadableLockMismatchException is a marker send from the EntryOffloadableSetUnlockOperation
-                    // meaning that the whole invocation of the EntryOffloadableOperation should be retried
-                    if (t instanceof EntryOffloadableLockMismatchException) {
-                        t = new RetryableHazelcastException(t.getMessage(), t);
+            private void retry(final Operation op) {
+                setUnlockRetryCount++;
+                if (isRetryLimitReached()) {
+                    try {
+                        Object exceeded = new HazelcastException("Set & Unlock retry limit exceeded for key " + dataKey);
+                        getOperationResponseHandler().sendResponse(EntryOperation.this, exceeded);
```

Comment:
Problem is, when you get a retryable response/exception (maybe apart from WrongTargetException, since operation is guaranteed to be local), you cannot give up until you get a non-retryable failure or a success response.

Suppose partition is migrating when you call unlock. So, you get retry exception and retried until you reach the cap. Then you give up and return failure response to the caller, without able to unlock. 
Then partition migration failed for some reason (target crashed etc), and migration rolled back. Now, partition is still on former node and key remains locked forever. 

Both migrations and heartbeat have timeout mechanisms. So, there should not be infinite blocking/retrying here. 

To me either there shouldn't be any retry limit or lock should be acquired with a lease (something greater than heartbeat timeout).